stats: build from source

4evy opened the pull request:
## Description of changes

@NixOS/darwin-maintainers Hey folks! I need a bit of helping hand, I think I *almost* have everything done for `stats` to be build from source but I'm struggling with an error that I have no idea how to fix

<details>
  <summary>Build log:</summary>
  <ul>
<li>

```console
$ nix-build -A stats
...
Running phase: unpackPhase
unpacking source archive /nix/store/az200p1ssqh25k1flyfrvcsfxq5sgh9d-source
calling 'unpackCmd' function hook '_defaultUnpack' /nix/store/az200p1ssqh25k1flyfrvcsfxq5sgh9d-source
source root is source
calling 'postUnpack' function hook '_updateSourceDateEpochFromSourceRoot'
Running phase: patchPhase
applying patch /nix/store/1f188qaadrz743a65n5y7lkfxhms81gc-0001-Fixup-Makefile.patch
patching file Makefile
evaling implicit 'postPatch' string hook
Running phase: updateAutotoolsGnuConfigScriptsPhase
Running phase: configurePhase
calling 'preConfigure' function hook '_multioutConfig'
no configure script, doing nothing
Running phase: buildPhase
build flags: SHELL=/nix/store/fyjay93q3dq2hx3dhx7zhr8kyjnkh9m8-bash-5.2p26/bin/bash
Actual version is: 582
Next version is: 583
Exporting application archive...
warning: destination option not implemented
warning: object version 54 may be unsupported
warning: unhandled Project key packageReferences
warning: unhandled ShellScriptBuildPhase key alwaysOutOfDate
warning: unhandled ShellScriptBuildPhase key inputFileListPaths
warning: unhandled ShellScriptBuildPhase key outputFileListPaths
warning: unhandled ShellScriptBuildPhase key alwaysOutOfDate
warning: unhandled ShellScriptBuildPhase key inputFileListPaths
warning: unhandled ShellScriptBuildPhase key outputFileListPaths
warning: unhandled NativeTarget key packageProductDependencies
warning: unhandled NativeTarget key packageProductDependencies
warning: unhandled NativeTarget key packageProductDependencies
Build settings from the command line:
    DERIVED_DATA_DIR=.
    SDKROOT=MacOSX11.0

=== BUILD TARGET Kit OF PROJECT Stats WITH CONFIGURATION Release ===

Check dependencies
warning: unable to find Swift libraries

Write auxiliary files
/bin/mkdir -p ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Intermediates/Stats.build/Release/Kit.build
write-file ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Intermediates/Stats.build/Release/Kit.build/Kit.hmap
write-file ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Intermediates/Stats.build/Release/Kit.build/Kit-own-target-headers.hmap
write-file ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Intermediates/Stats.build/Release/Kit.build/Kit-all-target-headers.hmap
write-file ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Intermediates/Stats.build/Release/Kit.build/Kit-all-non-framework-target-headers.hmap
write-file ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Intermediates/Stats.build/Release/Kit.build/Kit-generated-files.hmap
write-file ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Intermediates/Stats.build/Release/Kit.build/Kit-project-headers.hmap
write-file ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Intermediates/Stats.build/Release/Kit.build/module.modulemap
/bin/mkdir -p ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Intermediates/Stats.build/Release/Kit.build/Objects-normal/arm64
write-file ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Intermediates/Stats.build/Release/Kit.build/Objects-normal/arm64/Kit-OutputFileMap.json
write-file ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Intermediates/Stats.build/Release/Kit.build/unextended-module.modulemap
write-file ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Intermediates/Stats.build/Release/Kit.build/unextended-module-overlay.yaml
write-file ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Intermediates/Stats.build/Release/Kit.build/Objects-normal/arm64/Kit.LinkFileList

Create product structure

CpHeader ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Products/Release/Kit.bundle/Contents/PrivateHeaders/Kit.h '/private/tmp/nix-build-stats-2.11.5.drv-0/source/Kit/Supporting Files/Kit.h'
    cd /private/tmp/nix-build-stats-2.11.5.drv-0/source
    export DEVELOPER_DIR=/nix/store/60n185lykh89xmvmjl4nd6difsfa4bi3-xcodebuild-0.1.2-pre
    export PATH=/nix/store/kdiiaab1y56f0a6q3417wvb9ypl0wv2w-Platforms/MacOSX.platform/Developer/usr/bin:/nix/store/kdiiaab1y56f0a6q3417wvb9ypl0wv2w-Platforms/MacOSX.platform/usr/local/bin:/nix/store/kdiiaab1y56f0a6q3417wvb9ypl0wv2w-Platforms/MacOSX.platform/usr/bin:/nix/store/kdiiaab1y56f0a6q3417wvb9ypl0wv2w-Platforms/MacOSX.platform/usr/local/bin:/nix/store/3lbn1q3bbyxmfa59zwxk8g8952nl132p-Toolchains/XcodeDefault.xctoolchain/usr/bin:/nix/store/3lbn1q3bbyxmfa59zwxk8g8952nl132p-Toolchains/XcodeDefault.xctoolchain/usr/libexec:/nix/store/60n185lykh89xmvmjl4nd6difsfa4bi3-xcodebuild-0.1.2-pre/usr/bin:/nix/store/60n185lykh89xmvmjl4nd6difsfa4bi3-xcodebuild-0.1.2-pre/usr/local/bin:/nix/store/60n185lykh89xmvmjl4nd6difsfa4bi3-xcodebuild-0.1.2-pre/Tools:/nix/store/ncswzgmmm16xfckcrj9k57jx1yfp326y-ditto-impure-darwin/bin:/nix/store/ajjyja98l3xa05w6rvrzi14arj23g4sd-create-dmg-1.2.2/bin:/nix/store/60n185lykh89xmvmjl4nd6difsfa4bi3-xcodebuild-0.1.2-pre/bin:/nix/store/3lbn1q3bbyxmfa59zwxk8g8952nl132p-Toolchains/XcodeDefault.xctoolchain/bin:/nix/store/wsgvg07q93xmaqp6ln58kzr75rg6jm40-clang-wrapper-16.0.6/bin:/nix/store/xdw9w5y9xdjways9xzs7r02zdmky4bim-clang-16.0.6/bin:/nix/store/7k0qi2r54imwjfs2bklg7fv0mn5jglil-coreutils-9.5/bin:/nix/store/58jqw1w9kj3xm10jfgpr7n9rkvx3p2cy-cctools-binutils-darwin-wrapper-1010.6/bin:/nix/store/5nwhbvfxfhdz9192ihsjpqbsr5vvhbsk-cctools-binutils-darwin-1010.6/bin:/nix/store/y01yv3ylb4zyadnh8cmpv3h8j0vw9slx-findutils-4.10.0/bin:/nix/store/pj9qh3y52knawg34qfnk9nq1kk12lh48-diffutils-3.10/bin:/nix/store/0x1mwb2bl2mjm658kcciwyg7n5avzvfp-gnused-4.9/bin:/nix/store/5g9xsw7dyihfh1kcf73nqqrla3zyy3r0-gnugrep-3.11/bin:/nix/store/ssprxb26mz01akbhlswdb7nbc30iyjjh-gawk-5.2.2/bin:/nix/store/lzndf97akrpmd3vjdq06xzwk8fnjc4j7-gnutar-1.35/bin:/nix/store/ynhzyabgbx6fz49sy944ws9wnskangxc-gzip-1.13/bin:/nix/store/ci1mxirfvn961zh3m4w9mr513ii7s2lz-bzip2-1.0.8-bin/bin:/nix/store/6gylp4vygmsm12rafhzvklrfkbhwwq40-gnumake-4.4.1/bin:/nix/store/fyjay93q3dq2hx3dhx7zhr8kyjnkh9m8-bash-5.2p26/bin:/nix/store/ga67f01j6g0qq7y9flgf3fzr1lbdi58y-patch-2.7.6/bin:/nix/store/6rk472mwpc5zwfhh6siyycl28y2baqrf-xz-5.6.2-bin/bin:/nix/store/apqqsrdg6xsz9icx04siw3l4z9yrxws3-file-5.45/bin
    builtin-copy -resolve-src-symlinks -exclude .DS_Store -exclude CVS -exclude .svn -exclude .git -exclude .hg /private/tmp/nix-build-stats-2.11.5.drv-0/source/Kit/Supporting Files/Kit.h ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Products/Release/Kit.bundle/Contents/PrivateHeaders

Ditto ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Products/Release/Kit.bundle/Contents/Modules/Kit.swiftmodule/arm64.swiftmodule ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Intermediates/Stats.build/Release/Kit.build/Objects-normal/arm64/Kit.swiftmodule
    cd /private/tmp/nix-build-stats-2.11.5.drv-0/source
    /usr/bin/ditto -rsrc ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Intermediates/Stats.build/Release/Kit.build/Objects-normal/arm64/Kit.swiftmodule ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Products/Release/Kit.bundle/Contents/Modules/Kit.swiftmodule/arm64.swiftmodule
ditto: Cannot get the real path for source './Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Intermediates/Stats.build/Release/Kit.build/Objects-normal/arm64/Kit.swiftmodule'

** BUILD FAILED **

The following build commands failed:
    Ditto ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Products/Release/Kit.bundle/Contents/Modules/Kit.swiftmodule/arm64.swiftmodule ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Intermediates/Stats.build/Release/Kit.build/Objects-normal/arm64/Kit.swiftmodule
(1 failure)
make: *** [Makefile:17: archive] Error 1
```
</li>
  </ul>
</details>

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

matteo-pacini commented:
@DontEatOreo have you tried adding `swift` as a `nativeBuildInput`?

DontEatOreo commented:
> @DontEatOreo have you tried adding `swift` as a `nativeBuildInput`?

I will try it rn

DontEatOreo commented:
Unfortunately, that did not work *(I got same error)* **but** I added `swift swiftPackages.swift-docc swiftPackages.Foundation` to `nativeBuildInputs` and got a different error:

```console
./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Intermediates/Stats.build/Release/Kit.build/DerivedSources/Kit-Swift.h ./Stats-bkfkssqxiuvzqubnwpwicsycxaul/Build/Intermediates/Stats.build/Release/Kit.build/Objects-normal/arm64/Kit-Swift.h
```

I also added `export SWIFT_LIBRARY_PATH=${swift.swift.lib}/${swift.swiftModuleSubdir}` which got rid of `warning: unable to find Swift libraries`

matteo-pacini commented:
@DontEatOreo You may also want to go through the Swift files in the project, and check which frameworks are required: https://github.com/search?q=repo%3Aexelban%2Fstats%20%20path%3A*.swift%20%22import%22&type=code (`import` statements at the top)

I can see plenty, and I have a feeling they need to be added via `darwin.apple_sdk.frameworks`

DontEatOreo commented:
@matteo-pacini I went and added all of them, but still same issue, weirdly enough every time I try to build it gives different error even if nothing was changed once will be 

`Ditto ./Stats-errkzjlgwieyrecnhkbyzyvumguw/Build/Products/Release/Kit.bundle/Contents/Modules/Kit.swiftmodule/arm64.swiftmodule ./Stats-errkzjlgwieyrecnhkbyzyvumguw/Build/Intermediates/Stats.build/Release/Kit.build/Objects-normal/arm64/Kit.swiftmodule` 

another time will be `./Stats-errkzjlgwieyrecnhkbyzyvumguw/Build/Intermediates/Stats.build/Release/Kit.build/DerivedSources/Kit-Swift.h ./Stats-errkzjlgwieyrecnhkbyzyvumguw/Build/Intermediates/Stats.build/Release/Kit.build/Objects-normal/arm64/Kit-Swift.h` 

or `CompileC ./Stats-errkzjlgwieyrecnhkbyzyvumguw/Build/Intermediates/Stats.build/Release/Kit.build/Objects-normal/arm64/lldb.o Kit/lldb/lldb.m normal arm64 objective-c++ com.apple.compilers.llvm.clang.1_0.compiler`

These are the frameworks I added:

```
darwin.apple_sdk_11_0.frameworks.AppKit
darwin.apple_sdk_11_0.frameworks.Cocoa
darwin.apple_sdk_11_0.frameworks.CoreBluetooth
darwin.apple_sdk_11_0.frameworks.CoreWLAN
darwin.apple_sdk_11_0.frameworks.DeveloperToolsSupport
darwin.apple_sdk_11_0.frameworks.DiskArbitration
darwin.apple_sdk_11_0.frameworks.Foundation
darwin.apple_sdk_11_0.frameworks.IOBluetooth
darwin.apple_sdk_11_0.frameworks.IOKit
darwin.apple_sdk_11_0.frameworks.SensorKit
darwin.apple_sdk_11_0.frameworks.ServiceManagement
darwin.apple_sdk_11_0.frameworks.SwiftUI
darwin.apple_sdk_11_0.frameworks.SystemConfiguration
darwin.apple_sdk_11_0.frameworks.UserNotifications
darwin.apple_sdk_11_0.frameworks.WidgetKit
```

*(I tried `apple_sdk` too no difference...)*

Interestingly, even if I remove `ditto` from the Makefile, it still gets invoked...

matteo-pacini commented:
@DontEatOreo not sure what the problem is here, but a lot of Darwin apps I tried to compile from source failed because of `xcbuild`, which doesn't exactly work like Xcode's (https://github.com/facebookarchive/xcbuild, it's an archived project, last updated 5YA).

Could you please push your latest addition here so I can have a look? 

It would also be great to mark this PR as a draft for the time being.

DontEatOreo commented:
@matteo-pacini Yeah, that's unfortunate. AFAIK we can't really use the Apple Xcode build? I made the commands more "friendly" to Facebook's `xcbuild`, which doesn't have the `archive` feature but does have the `build"` one. It's also more "friendly" to Nix since I got rid of Ditto.

matteo-pacini commented:
@DontEatOreo Bit of a bummer indeed,  I had a similar issue here #320122, where I had to backtrack from source-based derivation to a bin-based one, as we cannot compile Metal shaders without Xcode (yet!).

Going to double check what happens later today, and come back to you!

matteo-pacini commented:
Tried the PR with a mix of changes (darwin frameworks, darwin.ditto also) - works in Xcode but fails on Nix :( perhaps someone from @NixOS/darwin-maintainers has a clue on how to fix this?

jwiegley commented:
I tried but was unable to build `swift-5.8` using this commit, which is a dependency.

DontEatOreo commented:
As far as I can tell *currently*, it's more or less impossible to build `stats` from source code, so I'm going to put this PR in draft, and hopefully in the future, it would be possible to build it from source.

I want to do a final ping to @NixOS/darwin-maintainers. If anyone has experience with Xcode, I would appreciate a review. The main issue is that `Ditto` always fails on either `Kit.swiftmodule`, `Kit-Swift.h`, or some other file. Every rebuild is a gamble on what error you get, even with no changes made to the file.

Running the build command outside Nix with Apple's XCode produces a `Stats.app` ready to be used.

n8henrie commented:
I apologize that I haven't had time to look much at this (moving this month and I'm relative novice anyway), but just wanted to mention my experience adapting my beloved `Notational Velocity.app` to build with nix recently. I had no luck using `xcbuild` in spite of days of attempts, inconsistent errors similar to what you mention above. I eventually just exported the build log when building with xcode (something like `xcodebuild clean; xcodebuild -configuration Deployment build |& tee xcbuild.log`) and used this to find the actual `clang` calls that were beneath the XCode magic and eventually got this to compile.

Not sure if that is applicable given the swift and swiftmodule issues noted above, but can't hurt to mention the (painstaking) approach that eventually worked for me.

https://github.com/n8henrie/nv/blob/master/default.nix

afh commented:
Maybe this gets easier when the [transition to new (darwin sdk) pattern](https://discourse.nixos.org/t/on-the-future-of-darwin-sdks-or-how-you-can-stop-worrying-and-put-the-sdk-in-build-inputs/50574) is complete?

emilazy commented on pkgs/by-name/st/stats/package.nix:
```diff
@@ -1,29 +1,44 @@
 
-  sourceRoot = ".";
+  nativeBuildInputs = [
+    apple-sdk
```
This is redundant; `apple-sdk` is part of the standard environment (and Swift propagates its required SDK anyway).

reckenrode replied:
SDKs also go in `buildInputs` not `nativeBuildInputs` (unless you’re using an SDK for a `depsBuildBuild` compiler, which isn’t the case here).

emilazy commented on pkgs/by-name/st/stats/package.nix:
```diff
@@ -1,29 +1,44 @@
 }:
 
-stdenvNoCC.mkDerivation (finalAttrs: {
+stdenv.mkDerivation (finalAttrs: {
```
You need to use `swiftPackages.stdenv` when using Swift; the build on `x86_64-darwin` is broken right now.

DontEatOreo commented:
https://gist.github.com/DontEatOreo/312885e0a4459e0e3e5ae34da2a11d3c

emilazy commented:
I wish I knew what was going on here. It seems like some sort of weird SDK issue and it sort of rhymes with problems I remember having when not using `swiftPackages.stdenv`, but I’m not sure exactly what's going on.

emilazy commented:
Playing around with this a bit and `<unknown>:0: error: underlying Objective-C module 'Kit' not found` seems suspicious. Do you know if running the equivalent commands to this derivation with an Xcode toolchain works? Especially an Xcode with Swift 5.8?

DontEatOreo commented:
Hey @emilazy, sorry for the late response! I downloaded Xcode 14.3 and tried running it, but I got an error:

```
In order to use “Xcode.app”, you need to update to the latest version.

The version of Xcode installed on this Mac is not compatible with macOS Sequoia. Download the latest version for free from the App Store.
```

I found this workaround, https://stackoverflow.com/a/78996420

This worked until I tried to open the project with Xcode when I got this error:

<details>

```console
2024-11-29 00:55:22.805 Xcode[70940:814742] ApplePersistence=NO
2024-11-29 00:55:23.807 Xcode[70940:814742] Regex scanner "(null)" shouldn't have a Tokenizer. Ignoring it.
2024-11-29 00:55:24.158 Xcode[70940:814742] DVTCoreDeviceEnabledState: DVTCoreDeviceEnabledState_Disabled set via user default (DVTEnableCoreDevice=disabled)
objc[70940]: Class _EntryGenericToGL is implemented in both /Applications/Xcode.app/Contents/SharedFrameworks/GPUToolsServices.framework/Versions/A/GPUToolsServices (0x14aa5d770) and /Applications/Xcode.app/Contents/PlugIns/GPUDebugger.ideplugin/Contents/Frameworks/GPUToolsAdvancedUI.framework/Versions/A/GPUToolsAdvancedUI (0x14928d1f0). One of the two will be used. Which one is undefined.
objc[70940]: Class _EntryGLToGeneric is implemented in both /Applications/Xcode.app/Contents/SharedFrameworks/GPUToolsServices.framework/Versions/A/GPUToolsServices (0x14aa5d7c0) and /Applications/Xcode.app/Contents/PlugIns/GPUDebugger.ideplugin/Contents/Frameworks/GPUToolsAdvancedUI.framework/Versions/A/GPUToolsAdvancedUI (0x14928d240). One of the two will be used. Which one is undefined.
objc[70940]: Class DYPixelFormatConverter is implemented in both /Applications/Xcode.app/Contents/SharedFrameworks/GPUToolsServices.framework/Versions/A/GPUToolsServices (0x14aa5d810) and /Applications/Xcode.app/Contents/PlugIns/GPUDebugger.ideplugin/Contents/Frameworks/GPUToolsAdvancedUI.framework/Versions/A/GPUToolsAdvancedUI (0x14928d290). One of the two will be used. Which one is undefined.
objc[70940]: Class ASVError is implemented in both /Applications/Xcode.app/Contents/SharedFrameworks/GPUToolsCore.framework/Versions/A/GPUToolsCore (0x13c180f90) and /Applications/Xcode.app/Contents/PlugIns/GPUDebugger.ideplugin/Contents/Frameworks/GPUToolsASVC.framework/Versions/A/GPUToolsASVC (0x139f9c320). One of the two will be used. Which one is undefined.
objc[70940]: Class DYGPUDerivedEncoderCounterInfo is implemented in both /Applications/Xcode.app/Contents/SharedFrameworks/GPUToolsShaderProfiler.framework/Versions/A/GPUToolsShaderProfiler (0x13c437e08) and /Applications/Xcode.app/Contents/PlugIns/GPUDebugger.ideplugin/Contents/Frameworks/GTShaderProfiler.framework/Versions/A/GTShaderProfiler (0x14a778338). One of the two will be used. Which one is undefined.
objc[70940]: Class DYGPUTimelineInfo is implemented in both /Applications/Xcode.app/Contents/SharedFrameworks/GPUToolsShaderProfiler.framework/Versions/A/GPUToolsShaderProfiler (0x13c437e58) and /Applications/Xcode.app/Contents/PlugIns/GPUDebugger.ideplugin/Contents/Frameworks/GTShaderProfiler.framework/Versions/A/GTShaderProfiler (0x14a778388). One of the two will be used. Which one is undefined.
objc[70940]: Class DYTimelineCounterGroup is implemented in both /Applications/Xcode.app/Contents/SharedFrameworks/GPUToolsShaderProfiler.framework/Versions/A/GPUToolsShaderProfiler (0x13c437ea8) and /Applications/Xcode.app/Contents/PlugIns/GPUDebugger.ideplugin/Contents/Frameworks/GTShaderProfiler.framework/Versions/A/GTShaderProfiler (0x14a7783d8). One of the two will be used. Which one is undefined.
objc[70940]: Class DYWorkloadGPUTimelineInfo is implemented in both /Applications/Xcode.app/Contents/SharedFrameworks/GPUToolsShaderProfiler.framework/Versions/A/GPUToolsShaderProfiler (0x13c437ef8) and /Applications/Xcode.app/Contents/PlugIns/GPUDebugger.ideplugin/Contents/Frameworks/GTShaderProfiler.framework/Versions/A/GTShaderProfiler (0x14a778428). One of the two will be used. Which one is undefined.
objc[70940]: Class GRCPerFrameDataClass is implemented in both /Applications/Xcode.app/Contents/SharedFrameworks/MTLToolsShaderProfiler.framework/Versions/A/MTLToolsShaderProfiler (0x13c4d57f0) and /Applications/Xcode.app/Contents/PlugIns/GPUDebugger.ideplugin/Contents/Frameworks/GTShaderProfiler.framework/Versions/A/GTShaderProfiler (0x14a778838). One of the two will be used. Which one is undefined.
objc[70940]: Class DYShaderAnalyzerNextGPU is implemented in both /Applications/Xcode.app/Contents/SharedFrameworks/MTLToolsShaderProfiler.framework/Versions/A/MTLToolsShaderProfiler (0x13c4d57a0) and /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/Library/GPUTools/PlugIns/GLToolsShaderProfilerMobileSupport.gtplugin/Contents/MacOS/GLToolsShaderProfilerMobileSupport (0x14c5e8600). One of the two will be used. Which one is undefined.
2024-11-29 00:55:27.531 Xcode[70940:814742] +[IMKClient subclass]: chose IMKClient_Legacy
2024-11-29 00:55:27.531 Xcode[70940:814742] +[IMKInputSession subclass]: chose IMKInputSession_Legacy
2024-11-29 00:55:29.076 Xcode[70940:814742] The class 'NSOpenPanel' overrides the method identifier.  This method is implemented by class 'NSWindow'
2024-11-29 00:55:36.941 Xcode[70940:814742] [MT] DVTAssertions: Warning in /System/Volumes/Data/SWE/Apps/DT/BuildRoots/BuildRoot7/ActiveBuildRoot/Library/Caches/com.apple.xbs/Sources/DVTFrameworks/DVTFrameworks-21779/DVTFoundation/FoundationClassCategories/DVTNSKeyValueObserving.swift:95
Details:  The block will not be called before the observed object is already invalid. This is because we are observing the 'valid' key without passing NSKeyValueObservingOptionPrior. When -invalidate is called it will call the block from -didChangeValueForKey: which is at the end of -invalidate. The solution is to pass the NSKeyValueObservingOptionPrior option when creating the token. self: <IDEWorkspace:0x161362700 path:'/Users/DontEatOreo/stats/Stats.xcodeproj/project.xcworkspace'>
Function: _dvt_warnIfKeyPathIsValidWithoutPrior(_:options:)
Thread:   <_NSMainThread: 0x60000135b680>{number = 1, name = main}
Please file a bug at https://feedbackassistant.apple.com with this warning message and any useful information you can provide.
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[IDEVariablesViewPopUpButtonCell _autolayout_cellSize]: unrecognized selector sent to instance 0x1613d8080'
*** First throw call stack:
(
	0   CoreFoundation                      0x0000000184c70ec0 __exceptionPreprocess + 176
	1   DVTFoundation                       0x0000000105e7bd84 DVTFailureHintExceptionPreprocessor + 392
	2   libobjc.A.dylib                     0x0000000184756cd8 objc_exception_throw + 88
	3   CoreFoundation                      0x0000000184d265a4 -[NSObject(NSObject) __retain_OA] + 0
	4   CoreFoundation                      0x0000000184bdd140 ___forwarding___ + 1568
	5   CoreFoundation                      0x0000000184bdca60 _CF_forwarding_prep_0 + 96
	6   IDEKit                              0x00000001094baaf4 -[IDEVariablesViewPopUpButtonCell cellSize] + 64
	7   IDEKit                              0x00000001094b5a18 -[IDEVariablesView _updateScopePopUpTitle] + 440
	8   IDEKit                              0x00000001094b3ea4 -[IDEVariablesView setContentProvider:] + 496
	9   IDEKit                              0x000000010951d6d8 -[IDEDefaultDebugArea viewDidInstall] + 292
	10  DVTViewControllerKit                0x0000000104977e24 -[DVTViewController _viewDidInstall] + 224
	11  DVTViewControllerKit                0x00000001049787a8 -[_DVTViewController_ViewLifecycleInterpositions viewDidMoveToWindow] + 40
	12  AppKit                              0x0000000188754a5c -[NSView _setWindow:] + 1800
	13  AppKit                              0x00000001891d0a3c __21-[NSView _setWindow:]_block_invoke.146 + 268
	14  CoreAutoLayout                      0x000000018dd33550 -[NSISEngine withBehaviors:performModifications:] + 88
	15  AppKit                              0x00000001894ae748 -[NSView(NSConstraintBasedLayoutInternal) _withAutomaticEngineOptimizationDisabled:] + 48
	16  AppKit                              0x0000000188754a54 -[NSView _setWindow:] + 1792
	17  AppKit                              0x000000018875b6c8 -[NSView addSubview:] + 212
	18  AppKit                              0x000000018887c014 -[NSView setSubviews:] + 448
	19  DVTUserInterfaceKit                 0x00000001078e09c0 -[DVTBorderedView setContentView:] + 164
	20  IDEKit                              0x00000001094d944c -[IDEDebugAreaBorderedView setContentView:] + 108
	21  IDEKit                              0x00000001094d93d8 __57-[IDEDebugAreaBorderedView _restoreContentViewToSubviews]_block_invoke + 28
	22  DVTFoundation                       0x0000000105db3be0 DVTInvokePreventingRecursion + 44
	23  IDEKit                              0x00000001094d93b0 -[IDEDebugAreaBorderedView _restoreContentViewToSubviews] + 84
	24  IDEKit                              0x00000001094d94bc -[IDEDebugAreaBorderedView ensureContentViewIsInHierarchyDuringBlock:] + 84
	25  IDEKit                              0x00000001094e0b74 -[IDEEditorArea revertState] + 104
	26  IDEKit                              0x00000001094dc864 __31-[IDEEditorArea viewDidInstall]_block_invoke + 340
	27  DVTFoundation                       0x0000000105e83af0 -[DVTObservingBlockToken observeValueForKeyPath:ofObject:change:context:] + 220
	28  Foundation                          0x0000000185d91f1c NSKeyValueNotifyObserver + 252
	29  Foundation                          0x0000000185e3ed34 NSKeyValueDidChange + 388
	30  Foundation                          0x0000000185d850b8 -[NSObject(NSKeyValueObservingPrivate) _changeValueForKeys:count:maybeOldValuesDict:maybeNewValuesDict:usingBlock:] + 736
	31  Foundation                          0x0000000185dada10 -[NSObject(NSKeyValueObservingPrivate) _changeValueForKey:key:key:usingBlock:] + 64
	32  Foundation                          0x0000000185e6cd14 _NSSetBoolValueAndNotify + 408
	33  IDEFoundation                       0x000000010abb54e8 __81-[IDEWorkspace _finishLoadingAsynchronously:shouldUpgradeFromSimpleFilesFocused:]_block_invoke + 240
	34  IDEFoundation                       0x000000010abb56b0 __81-[IDEWorkspace _finishLoadingAsynchronously:shouldUpgradeFromSimpleFilesFocused:]_block_invoke.142 + 84
	35  Foundation                          0x0000000185d9eab0 __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__ + 24
	36  Foundation                          0x0000000185d9e970 -[NSBlockOperation main] + 104
	37  Foundation                          0x0000000185d9e900 __NSOPERATION_IS_INVOKING_MAIN__ + 16
	38  Foundation                          0x0000000185d9dc60 -[NSOperation start] + 648
	39  Foundation                          0x0000000185d9d9d0 __NSOPERATIONQUEUE_IS_STARTING_AN_OPERATION__ + 16
	40  Foundation                          0x0000000185d9d8c0 __NSOQSchedule_f + 172
	41  libdispatch.dylib                   0x0000000184973870 _dispatch_block_async_invoke2 + 148
	42  libdispatch.dylib                   0x0000000184964658 _dispatch_client_callout + 20
	43  libdispatch.dylib                   0x0000000184972f68 _dispatch_main_queue_drain + 980
	44  libdispatch.dylib                   0x0000000184972b84 _dispatch_main_queue_callback_4CF + 44
	45  CoreFoundation                      0x0000000184c3d680 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 16
	46  CoreFoundation                      0x0000000184bfd17c __CFRunLoopRun + 1996
	47  CoreFoundation                      0x0000000184bfc334 CFRunLoopRunSpecific + 572
	48  HIToolbox                           0x00000001900340cc RunCurrentEventLoopInMode + 292
	49  HIToolbox                           0x0000000190039ebc ReceiveNextEventCommon + 636
	50  HIToolbox                           0x000000019003a020 _BlockUntilNextEventMatchingListInModeWithFilter + 76
	51  AppKit                              0x000000018873fa70 _DPSNextEvent + 660
	52  AppKit                              0x00000001890657b8 -[NSApplication(NSEventRouting) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 688
	53  DVTKit                              0x0000000104e24ac8 -[DVTApplication nextEventMatchingMask:untilDate:inMode:dequeue:] + 300
	54  AppKit                              0x0000000188732b7c -[NSApplication run] + 480
	55  DVTKit                              0x0000000104e23cd0 -[DVTApplication run] + 60
	56  AppKit                              0x000000018870944c NSApplicationMain + 888
	57  dyld                                0x0000000184794274 start + 2840
)
2024-11-29 00:55:36.982 Xcode[70940:814742] [MT] DVTAssertions: UNCAUGHT EXCEPTION (NSInvalidArgumentException): -[IDEVariablesViewPopUpButtonCell _autolayout_cellSize]: unrecognized selector sent to instance 0x1613d8080
UserInfo: (null)
Open FDs: 47/7168
Hints: 
2024-11-29 00:55:37.075 Xcode[70940:814742] [MT] DVTAssertions: 
Backtrace:
  0   __exceptionPreprocess (in CoreFoundation)
  1   DVTFailureHintExceptionPreprocessor (in DVTFoundation)
  2   objc_exception_throw (in libobjc.A.dylib)
  3   -[NSObject(NSObject) __retain_OA] (in CoreFoundation)
  4   ___forwarding___ (in CoreFoundation)
  5   _CF_forwarding_prep_0 (in CoreFoundation)
  6   -[IDEVariablesViewPopUpButtonCell cellSize] (in IDEKit)
  7   -[IDEVariablesView _updateScopePopUpTitle] (in IDEKit)
  8   -[IDEVariablesView setContentProvider:] (in IDEKit)
  9   -[IDEDefaultDebugArea viewDidInstall] (in IDEKit)
 10   -[DVTViewController _viewDidInstall] (in DVTViewControllerKit)
 11   -[_DVTViewController_ViewLifecycleInterpositions viewDidMoveToWindow] (in DVTViewControllerKit)
 12   -[NSView _setWindow:] (in AppKit)
 13   __21-[NSView _setWindow:]_block_invoke.146 (in AppKit)
 14   -[NSISEngine withBehaviors:performModifications:] (in CoreAutoLayout)
 15   -[NSView(NSConstraintBasedLayoutInternal) _withAutomaticEngineOptimizationDisabled:] (in AppKit)
 16   -[NSView _setWindow:] (in AppKit)
 17   -[NSView addSubview:] (in AppKit)
 18   -[NSView setSubviews:] (in AppKit)
 19   -[DVTBorderedView setContentView:] (in DVTUserInterfaceKit)
 20   -[IDEDebugAreaBorderedView setContentView:] (in IDEKit)
 21   __57-[IDEDebugAreaBorderedView _restoreContentViewToSubviews]_block_invoke (in IDEKit)
 22   DVTInvokePreventingRecursion (in DVTFoundation)
 23   -[IDEDebugAreaBorderedView _restoreContentViewToSubviews] (in IDEKit)
 24   -[IDEDebugAreaBorderedView ensureContentViewIsInHierarchyDuringBlock:] (in IDEKit)
 25   -[IDEEditorArea revertState] (in IDEKit)
 26   __31-[IDEEditorArea viewDidInstall]_block_invoke (in IDEKit)
 27   -[DVTObservingBlockToken observeValueForKeyPath:ofObject:change:context:] (in DVTFoundation)
 28   NSKeyValueNotifyObserver (in Foundation)
 29   NSKeyValueDidChange (in Foundation)
 30   -[NSObject(NSKeyValueObservingPrivate) _changeValueForKeys:count:maybeOldValuesDict:maybeNewValuesDict:usingBlock:] (in Foundation)
 31   -[NSObject(NSKeyValueObservingPrivate) _changeValueForKey:key:key:usingBlock:] (in Foundation)
 32   _NSSetBoolValueAndNotify (in Foundation)
 33   __81-[IDEWorkspace _finishLoadingAsynchronously:shouldUpgradeFromSimpleFilesFocused:]_block_invoke (in IDEFoundation)
 34   __81-[IDEWorkspace _finishLoadingAsynchronously:shouldUpgradeFromSimpleFilesFocused:]_block_invoke.142 (in IDEFoundation)
 35   __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__ (in Foundation)
 36   -[NSBlockOperation main] (in Foundation)
 37   __NSOPERATION_IS_INVOKING_MAIN__ (in Foundation)
 38   -[NSOperation start] (in Foundation)
 39   __NSOPERATIONQUEUE_IS_STARTING_AN_OPERATION__ (in Foundation)
 40   __NSOQSchedule_f (in Foundation)
 41   _dispatch_block_async_invoke2 (in libdispatch.dylib)
 42   _dispatch_client_callout (in libdispatch.dylib)
 43   _dispatch_main_queue_drain (in libdispatch.dylib)
 44   _dispatch_main_queue_callback_4CF (in libdispatch.dylib)
 45   __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ (in CoreFoundation)
 46   __CFRunLoopRun (in CoreFoundation)
 47   CFRunLoopRunSpecific (in CoreFoundation)
 48   RunCurrentEventLoopInMode (in HIToolbox)
 49   ReceiveNextEventCommon (in HIToolbox)
 50   _BlockUntilNextEventMatchingListInModeWithFilter (in HIToolbox)
 51   _DPSNextEvent (in AppKit)
 52   -[NSApplication(NSEventRouting) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] (in AppKit)
 53   -[DVTApplication nextEventMatchingMask:untilDate:inMode:dequeue:] (in DVTKit)
 54   -[NSApplication run] (in AppKit)
 55   -[DVTApplication run] (in DVTKit)
 56   NSApplicationMain (in AppKit)
 57   start (in dyld)
[1]    70940 abort      /Applications/Xcode.app/Contents/MacOS/Xcode
```

</details>

So it seems that I can't really test compiling with older Xcode on my end...

reckenrode commented:
I noticed this in the log.
```
    "-I",
    "macosx\/usr\/lib\/swift",
```
xcbuild uses a symbolic SDK name because it can’t handle an absolute path to an SDK, which results in an incorrect `SDKROOT` for the wrapper. The Swift wrapper needs updating to use `$DEVELOPER_DIR/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk` instead of `SDKROOT`. Unless someone beats me to it, I’ll open a PR in the morning (EST) to fix it. That just leaves the `error: underlying Objective-C module 'Kit' not found` error.

DontEatOreo commented:
Hey @reckenrode, thanks a lot for taking the time to review my PR. It seems like my PR has helped spot an issue with the Swift wrapper.

If we're only left with `error: underlying Objective-C module 'Kit' not found`, what can be done about it?

reckenrode commented:
https://github.com/NixOS/nixpkgs/pull/360207 is the PR for the Swift fix. I’m not sure what’s happening with `Kit`. It’s supposed to be built as part of the build, isn’t it?

DontEatOreo commented:
I rebased on top of master and cherry-picked c578c48b7f0bdecaf6c6eb6cef4678f2223b29a9, and it seems like #360207 is working, but building stats is still bogged down by other stuff too...

Here's a build log of `stats` with c578c48b7f0bdecaf6c6eb6cef4678f2223b29a9 cherry picked:
https://gist.githubusercontent.com/DontEatOreo/9876a59f5e2be5c94475b7809679ffd3/raw/5262a3d9299e06305916bf7a44ad9b8ba895d6a2/2024-12-28-building-stats-with-pr-360207

4evy committed:
stats: build from source

Also update to latest version:

Changelog: https://github.com/exelban/stats/releases/tag/v2.12.7
Diff: https://github.com/exelban/stats/compare/v2.12.1...v2.12.7

viraptor commented on pkgs/by-name/st/stats/package.nix:
```diff
@@ -1,29 +1,372 @@
+    # CWPHYMode.mode11be (WiFi 7) requires macOS 15+ SDK; @unknown default covers it
+    sed -i '/mode11be/d' Modules/Net/readers.swift
+
+    # Flatten nested xcassets groups (actool reimplementation doesn't handle subdirectories)
```
It will in just a moment. Please stand by...

viraptor replied:
Implemented: https://github.com/NixOS/nixpkgs/pull/504838

FlameFlag replied:
Thanks a lot!

viraptor reviewed (commented):
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 333546`
Commit: `0ea15702cd067e8f83e20d304821ca91cac10f4c`

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>stats</li>
  </ul>
</details>

Builds and works for me now.
